reward/stakingInfoCache.go

```diff
@@ -0,0 +1,83 @@
+// Copyright 2019 The klaytn Authors
+// This file is part of the klaytn library.
+//
+// The klaytn library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The klaytn library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the klaytn library. If not, see <http://www.gnu.org/licenses/>.
+
+package reward
+
+import "sync"
+
+const (
+	maxStakingCache = 4
+)
+
+type stakingInfoCache struct {
+	cells       map[uint64]*StakingInfo
+	minBlockNum uint64
+	lock        sync.RWMutex
+}
+
+func newStakingInfoCache() *stakingInfoCache {
+	stakingCache := new(stakingInfoCache)
+	stakingCache.cells = make(map[uint64]*StakingInfo)
+	return stakingCache
+}
+
+func (sc *stakingInfoCache) get(blockNum uint64) *StakingInfo {
+	sc.lock.RLock()
+	defer sc.lock.RUnlock()
+
+	if s, ok := sc.cells[blockNum]; ok {
+		return s
+	}
+	return nil
+}
+
+func (sc *stakingInfoCache) add(stakingInfo *StakingInfo) {
+	sc.lock.Lock()
+	defer sc.lock.Unlock()
+
+	// Assumption: stakingInfo should not be nil.
+
+	if _, ok := sc.cells[stakingInfo.BlockNum]; ok {
+		return
+	}
+
+	if len(sc.cells) < maxStakingCache {
+		// empty room available
+		sc.cells[stakingInfo.BlockNum] = stakingInfo
+		if stakingInfo.BlockNum < sc.minBlockNum || len(sc.cells) == 1 {
+			// new minBlockNum or newly inserted one is the first element
+			sc.minBlockNum = stakingInfo.BlockNum
+		}
+		return
+	}
+
+	// evict one and insert new one
+	delete(sc.cells, sc.minBlockNum)
+
+	// update minBlockNum
+	if stakingInfo.BlockNum < sc.minBlockNum {
+		sc.minBlockNum = stakingInfo.BlockNum
+	} else {
+		min := stakingInfo.BlockNum
+		for _, s := range sc.cells {
```

Comment:
Once the max is full, we have to do a linear search every time. Is this logic without performance issues?

Reply:
well, if the cache size is big, it would have performance issues.
but, i think it would be okay because the maximum of the cache is 4.